--[[

				_______________________________________________________________________________________________

											  __ \    ____|   ____|  _ _|    \  |
											  |   |   __|     |        |      \ |
											  |   |   |       __|      |    |\  |
											 ____/   _____|  _|      ___|  _| \_|

				________________________________ THE CUSTOM INSTANCE FRAMEWORK ________________________________


				Defin is a light user-made framework that allows you to create custom instances before run-time.
				By simply creating a script module under the Classes folder, you can then enter properties,
				constants, client/server/request/replicate methods and events that are automatically fired
				during the lifecycle of an instance.

				_______________________________________________________________________________________________


				NOTE:

						This is the first release candidate for Defin. It severely lacks documentation and
						technical support. Use at your own risk.

				_______________________________________________________________________________________________


				FEATURES:

						-	Custom instances
						-	Custom components implementation
						-	Custom services
						- 	Automatically replicated properties
						- 	Class constants
						- 	Client / Server / Client -> Server -> Client / Server -> Client(s) methods
						- 	Server & Client lifecycle hooks on Initialized, Property changed and Destroyed
						- 	Secure Client / Server separation
						- 	Default method compatibility for Destroy, Clone and IsA
						-	Anti game theft / manipulation
						-	Organised structure

				_______________________________________________________________________________________________


				Setup:				By pressing the "Setup" button, the starting template source for a Defin
									project will be put under ServerStorage. This template is strictly used
									for development and will be omitted on run-time.

				Generate: 			By pressing the "Generate" button, your development template source will be
									structured for run-time. Automatically, the content will be put under
									ReplicatedStorage, the server initializer will be put under
									ServerScriptService and the client initializer will be put under
									StarterPlayerScripts. Your project will now be ready for run-time.

				Minify (optional):	By pressing the "Minify" button, your game structure that was generated by
									using the previous button will be minified. This is entirely optional,
									however it is recommended to use it before publishing your game. Minifying
									your scripts allows you to considerably lower their size and also make them
									very difficult for exploiters to figure out.

				Clean (optional):	By pressing the "Clean" button, your generated game structure will be
									deleted. Note that this procedure is executed when generating your project.


--]]

local DEFIN_LOCKED		= "DefinLocked"

local defin							= {
	Nil = "DefinNil"
}
local containers					= {}
local instances						= {}
local serviceInstances				= {}
local instanceThreads				= {}
local services						= {}
local classes						= {}
local components					= {}
local classesWithComponent			= {}
local replicationBundles			= {}
local pausedInstanceReplications	= {}
local playerRequests				= {}
local plugins						= {}
local tablesForModuleTypes			= {
	Classes = classes,
	Components = components,
	Services = services,
	Plugins = plugins
}

local function newClassTable()
	return {
		Server = { Changed = {}, Request = {}, Thread = {} },
		Client = { Changed = {}, Replicate = {}, Thread = {} },
		Components = {},
		Constants = {},
		Properties = {},
		Changed = {},
		Thread = {},
		ContainerClassName = "Folder"
	}
end

defin.Classes = setmetatable({}, {
	__index = function(_, name)
		if not classes[name] then
			classes[name] = newClassTable()
		end
		return classes[name]
	end,
	__newindex = function(_, name, value)
		classes[name] = value
	end,
})

defin.Components = setmetatable({}, {
	__index = function(_, name)
		if not components[name] then
			components[name] = newClassTable()
		end
		return components[name]
	end,
	__newindex = function(_, name, value)
		components[name] = value
	end,
})

local runService					= game:GetService("RunService")
local collectionService 			= game:GetService("CollectionService")
local replicatedStorage				= game:GetService("ReplicatedStorage")
local serverStorage					= game:GetService("ServerStorage")
local serverScriptService				= game:GetService("ServerScriptService")
local players					= game:GetService("Players")
local localPlayer						= players.LocalPlayer
local isClient						= runService:IsClient()
local isServer						= runService:IsServer()
local isStudio						= runService:IsStudio()
local context						= nil
local eventsFolder
local definMetatable				= nil
local DEBUG                         = isStudio

--[[

____ FUNCTION - NEW_INSTANCE _______________________________________________________________________________________________________________________________________________

Creates and returns a new Defin instance

className:	Name of the class to be instanced
parent (optional):		Parent of the instance
container (optional):	ROBLOX instance that will be wrapped by the custom instance

--]]

function defin:NewInstance(className, parent, container)
	local class, instance
	
	if type(className) == "table" then
		instance = className
		container = instance["1"]
		className = instance["2"]
		
		if defin:IsWrapped(instance) then
			return instance
		else
			class = classes[className] or services[className]
		end
	else
		class = classes[className] or services[className]
		instance = {}
		
		if not class then
			error('Unable to create a Defin Instance of type "' .. className .. '"')
		end
		
		for index, value in pairs(class.PropertyIndexes) do
			instance[value] = defin:CloneValue(class.Properties[index])
		end
		
		if containers[className] or containers[container] then
			if typeof(container) ~= "Instance" then
				if container then
					container = containers[container]
				else
					container = containers[className]
				end

				if services[className] then
					container.Parent = parent
				else
					container = container:Clone()
				end
			elseif not container:FindFirstChild("PackageLink") then
				for _, child in pairs(containers[className]:GetChildren()) do
					if not container:FindFirstChild(child.Name) and not child:IsA("PackageLink") then
						child:Clone().Parent = container
					end
				end
			end
		else
			container = container or Instance.new(class.ContainerClassName or "Folder")
		end
		
		instance["1"] = container
		instance["2"] = className
		instance["3"] = instance["3"] or tostring(instance):sub(8)
		instance["4"] = services[className] ~= nil

		if not container.Parent then
			container.Name = className
			container.Parent = defin:Unwrap(parent)
		end
		
		collectionService:AddTag(container, "Defin")
	end
	
	container.AncestryChanged:Connect(function()
		if not instance:IsDescendantOf(game) then
			destroyInstance(instance)
		end
	end)
	
	instance = setmetatable(instance, definMetatable)

	initializeInstance(instance)

	if isServer then
		if not checkReplicationBundle(instance) then
			replicateInstance(instance)
		end
		
		instance:GetPropertyChangedSignal("Parent"):Connect(function()
			if not checkReplicationBundle(instance) then
				replicateInstance(instance)
			end
		end)
	end
	
	return instance
end

--[[
	
____ FUNCTION - IS_WRAPPED _________________________________________________________________________________________________________________________________________________

Returns whether or not a value is wrapped into a Defin instance

value: The value

--]]

function defin:IsWrapped(value)
	return getmetatable(value) == DEFIN_LOCKED
end

--[[
	
____ FUNCTION - WRAP _______________________________________________________________________________________________________________________________________________________

Returns the wrapped Defin instance of a value or the value itself

value: The value to be wrapped

--]]

function defin:Wrap(value)
	return instances[value] or value
end

--[[

____ FUNCTION - UNWRAP _____________________________________________________________________________________________________________________________________________________

Returns the unwrapped value of a value

value: The value to be unwrapped

--]]

function defin:Unwrap(value)
	if defin:IsWrapped(value) then
		return value["1"]
	else
		return value
	end
end

--[[

____ FUNCTION - WAIT_FOR_WRAP ______________________________________________________________________________________________________________________________________________

Yields for the value to be wrapped then returns it

value: The value to wait to be wrapped

--]]

function defin:WaitForWrap(value, timeOut)
	if value then
		local start = tick()
		local warned = false

		while true do
			local wrappedValue = defin:Wrap(value)

			if defin:IsWrapped(wrappedValue) then
				return wrappedValue
			elseif typeof(value) == "Instance" and collectionService:HasTag(value, "Defin") then
				if timeOut and tick() - start > timeOut then
					return
				elseif not warned and not timeOut and tick() - start > 5 then
					warn("Infinite yield possible on 'Defin:WaitForWrap(" .. value.Name .. ")'")
					
					warned = true
				end

				runService.Heartbeat:Wait()
			else
				return
			end
		end
	end
end

--[[

____ FUNCTION - GET_SERVICE ______________________________________________________________________________________________________________________________________________

Yields for the service to be initialized then returns it

serviceName: The name of the service
ignoreInitialization: Ignores whether or not the server has finished initializing (cheap solution to avoiding cyclic dependency)

--]]

function defin:GetService(serviceName, ignoreInitialization)
	if not serviceInstances[serviceName] and not ignoreInitialization then
		while not serviceInstances[serviceName] do
			runService.Heartbeat:Wait()
		end
	end
	
	return serviceInstances[serviceName]
end

--[[

____ FUNCTION - GET_PLUGIN ______________________________________________________________________________________________________________________________________________

Plugins can be added inside the Plugins folder

pluginName: The name of the plugin

--]]

function defin:GetPlugin(pluginName)
	return plugins[pluginName]
end

function defin:PauseInstanceReplication(instance)
	if not isInstanceReplicationPaused(instance) then
		pausedInstanceReplications[instance] = {}
	end
end

function defin:ContinueInstanceReplication(instance)
	local updates = pausedInstanceReplications[instance]

	if updates then
		pausedInstanceReplications[instance] = nil

		if #updates > 0 then
			eventsFolder.InstanceReplicationContinued:FireAllClients(defin:Unwrap(instance), updates)
		end
	end
end

function defin:BindToReplicationBundle(classes)
	for _, class in pairs(classes) do
		if not replicationBundles[class] then
			replicationBundles[class] = {}
		end
	end
end

function defin:UnbindFromReplicationBundle(classes)
	for _, class in pairs(classes) do
		local _instances = replicationBundles[class]

		if _instances and #_instances > 0 then
			for _, instance in pairs(_instances) do
				instance["4"] = true
			end

			eventsFolder.InstancesAdded:FireAllClients(_instances)
		end

		replicationBundles[class] = nil
	end
end

function defin:GetRawContainer(className)
	return containers[className]
end

function defin:GetContainer(className)
	return containers[className]:Clone()
end

function defin:GetPlayer()
	local player = defin:Wrap(localPlayer)
	
	if player and not defin:IsWrapped(player) then
		player = defin:WaitForWrap(player)
	end
	
	return player
end

function defin:IsAClass(className)
	return classes[className] ~= nil
end

function defin:IsAClassFunction(classNameOrInstance, functionName)
	return type(classes[defin:IsWrapped(classNameOrInstance) and classNameOrInstance.ClassName or classNameOrInstance][functionName]) == "function"
end

function defin:IsClient()
	return isClient
end

function defin:IsServer()
	return isServer
end

function defin:IsStudio()
	return isStudio
end

function defin:GetConstants(className)
	return (classes[className] or components[className]).Constants
end

function defin:GetProperties(className)
	return (classes[className] or components[className]).Properties
end

function defin:ClassHasComponent(className, componentName)
	local class = classes[className]

	if class then
		if componentName then
			return table.find(class.Components, componentName) ~= nil
		else
			return #class.Components > 0
		end
	end
end

function defin:ClassHasAnyComponent(className, componentNames)
	for _, componentName in pairs(componentNames) do
		if defin:ClassHasComponent(className, componentName) then
			return true
		end
	end
	
	return false
end

function defin:GetClassesWithComponent(componentName)
	if classesWithComponent[componentName] then
		return classesWithComponent[componentName]
	else
		local classNames = {}

		for className, class in pairs(classes) do
			if table.find(class.Components, componentName) then
				table.insert(classNames, className)
			end
		end

		classesWithComponent[componentName] = classNames

		return classNames
	end
end

function defin:GetClassesWithAnyComponent(componentNames)
	local classNames = {}

	for className, class in pairs(classes) do
		for _, component in ipairs(componentNames) do
			if table.find(class.Components, component) then
				table.insert(classNames, className)
				break
			end
		end
	end

	return classNames
end

function defin:GetClassComponents(className)
	return classes[className].Components
end

function defin:GetComponents()
	local names = {}
	for name, _ in pairs(components) do
		table.insert(names, name)
	end
	return names
end

function defin:WrapTable(_table, waitForWrap)
	if type(_table) == "table" then
		local wrappedTable = {}
		for index, value in pairs(_table) do
			if waitForWrap and typeof(value) == "Instance" and collectionService:HasTag(value, "Defin") then
				wrappedTable[index] = defin:WaitForWrap(value, 10)
			else
				wrappedTable[index] = defin:Wrap(value)
			end
		end
		return wrappedTable
	end
end

function defin:UnwrapTable(_table)
	if type(_table) == "table" then
		local unwrappedTable = {}
		
		for index, value in pairs(_table) do
			unwrappedTable[index] = defin:Unwrap(value)
		end
		
		return unwrappedTable
	end
end

function defin:CloneValue(value)
	if type(value) == "table" then
		local _table = {}
		
		for index, _value in pairs(value) do
			_table[index] = defin:CloneValue(_value)
		end
		
		return _table
	elseif typeof(value) == "Instance" then
		local clone = value:Clone()
		clone.Parent = value.Parent
		
		return clone
	else
		return value
	end
end

function startInstanceThread(instance, name)
	if not isInstanceThreadActive(instance, name) then
		local class = getClassForInstance(instance)
		local func = class.Thread[name]

		if func then
			local container = instance.Container
			local _instanceThreads = instanceThreads[container]

			if not _instanceThreads then
				_instanceThreads = {}

				instanceThreads[container] = _instanceThreads
			end

			table.insert(_instanceThreads, name)

			coroutine.wrap(function()
				local delta = 0
				local _delay, paramA, paramB, paramC

				while instances[container] and isInstanceThreadActive(instance, name) do
					_delay, paramA, paramB, paramC = func(instance, delta, paramA, paramB, paramC)

					local _tick = tick()

					task.wait(_delay or 1)

					delta = tick() - _tick
				end
			end)()
		end
	end
end

function stopInstanceThread(instance, name)
	if isInstanceThreadActive(instance, name) then
		local container = instance.Container
		local index = table.find(instanceThreads[container], name)

		table.remove(instanceThreads[container], index)
	end
end

function isInstanceThreadActive(instance, name)
	local container = instance.Container
	local _instanceThreads = instanceThreads[container]

	if _instanceThreads then
		return table.find(_instanceThreads, name) ~= nil
	else
		return false
	end
end

--[[

____ INTERNAL FUNCTIONS ___________________________________________________________________________________________________________________________________________________


We recommend not making any changes

--]]

function getDefin()
	return defin, {
		Server = {
			Changed = {},
			Request = {},
			Thread = {}
		},
		Client = {
			Changed = {},
			Replicate = {},
			Thread = {}
		},
		Components = {},
		Constants = {},
		Properties = {},
		Changed = {},
		Thread = {},
		ContainerClassName = "Folder"
	}
end

definMetatable = {
	__metatable = DEFIN_LOCKED,

	__tostring = function(instance)
		return instance.Name .. " (" .. instance.ClassName .. ")"
	end,
	
	__index = function(instance, index)
		local class = getClassForInstance(instance)

		if index == "Container" then
			return instance["1"]
		elseif index == "ClassName" then
			return instance["2"]
		elseif index == "Id" then
			return instance["3"]
		elseif class.PropertyIndexes[index] then
			return defin:Wrap(rawget(instance, class.PropertyIndexes[index]))
		elseif class.ConstantIndexes[index] then
			return class.Constants[index]
		elseif type(class[index]) == "function" then
			return class[index]
		elseif type(instance["1"][index]) == "function" then
			return function(_, ...)
				return defin:Wrap(instance["1"][index](instance["1"], ...))
			end
		else
			return defin:Wrap(instance["1"][index])
		end
	end,
		
	__newindex = function(instance, index, newValue)
		local class = getClassForInstance(instance)
		local propertyIndex = class.PropertyIndexes[index]

		newValue = defin:Unwrap(newValue)
		
		if propertyIndex then
			local changed

			changed = changeInstanceProperty(instance, index, newValue, false, false)

			if changed and canReplicateInstance(instance) then
				eventsFolder.InstancePropertyFunction:FireAllClients(instance["1"], index, newValue)
			elseif isInstanceReplicationPaused(instance) then
				queueInstanceReplicationUpdate(instance, {
					1,
					index,
					newValue
				})
			end
		else
			instance["1"][index] = newValue
		end
	end
}

function initializeInstance(instance)
	if not instances[instance.Container] then
		local class = getClassForInstance(instance)
		
		instances[instance.Container] = instance
		
		for _, value in pairs(instance) do
			defin:WaitForWrap(value, 10)
		end
		
		for _, componentName in pairs(class.Components) do
			local func = class.InitializedFunctions[componentName]
			
			if type(func) == "function" then
				func(instance)
			end
		end
		
		local func = class.InitializedFunctions.Class
		
		if type(func) == "function" then
			func(instance)
		end
		
		for index in pairs(class.PropertyIndexes) do
			changeInstanceProperty(instance, index, instance[index], true, false)
		end
	end
end

function changeInstanceProperty(instance, propertyName, newValue, initializing, replicating)
	local class = getClassForInstance(instance)
	local index = class.PropertyIndexes[propertyName]
	local oldValue = rawget(instance, index)
	
	if (newValue ~= oldValue or initializing) and newValue ~= instance.Container then
		rawset(instance, index, newValue)
		
		local funcs = class.ChangedFunctions[propertyName]
		
		if funcs then
			oldValue = defin:Wrap(oldValue)
			newValue = defin:Wrap(newValue)
			
			for _, componentName in pairs(class.Components) do
				local func = funcs[componentName]
				
				if type(func) == "function" then
					func(instance, newValue, oldValue, initializing, replicating)
				end
			end
			
			local func = funcs.Class
			
			if type(func) == "function" then
				func(instance, newValue, oldValue, initializing, replicating)
			end
		end
		
		return true
	else
		return false
	end
end

function destroyInstance(instance)
	local class = getClassForInstance(instance)
	local container = instance.Container
	
	if instances[container] then
		instanceThreads[container] = nil
		
		for _, componentName in pairs(class.Components) do
			local func = class.DestroyedFunctions[componentName]
			
			if type(func) == "function" then
				if isStudio then
					func(instance)
				else
					pcall(function()
						func(instance)
					end)
				end
			end
		end
		
		local func = class.DestroyedFunctions.Class
		
		if type(func) == "function" then
			if isStudio then
				func(instance)
			else
				pcall(function()
					func(instance)
				end)
			end
		end
		
		instances[container] = nil
		
		container:Destroy()
	end
end

function queueInstanceReplicationUpdate(instance, update)
	local updates = pausedInstanceReplications[instance]

	if updates then
		table.insert(updates, update)
	end
end

function getClassForInstance(instance)
	return classes[instance["2"]] or services[instance["2"]]
end

function requireModule(_table, moduleScript)
	local moduleName = moduleScript.Name
	local module = require(moduleScript)
	
	_table[moduleName] = module
	
	if isClient then
		hideInstance(moduleScript)
	end
end

function setupClassComponents(class, component)
	component = component or class

	for _, componentName in pairs(component.Components) do
		if not table.find(class.Components, componentName) then
			table.insert(class.Components, componentName)
		end

		local _component = components[componentName]

		setupClassComponents(class, _component)
	end
end

function setupModule(moduleName, moduleType, module)
	if moduleType == "Class" or moduleType == "Service" then
		module.InitializedFunctions = {}
		module.ChangedFunctions = {}
		module.DestroyedFunctions = {}

		setupClassComponents(module)
	
		for _, componentName in pairs(module.Components) do
			local component = components[componentName]
			
			if not component then
				error('"' .. componentName .. '" is not a valid component for class ' .. moduleName)
			end

			if not module.ContainerClassName and type(component.ContainerClassName) == "string" then
				module.ContainerClassName = component.ContainerClassName
			end
			
			for index, value in pairs(component.Constants) do
				if type(index) == "string" then
					if module.Constants[index] == nil  then
						if value == defin.Nil then
							value = nil
						end
						module.Constants[index] = value
					elseif type(value) == "table" and type(module.Constants[index]) == "table" then
						if isArray(value) and isArray(module.Constants[index]) then
							if not module.Constants[index] then
								module.Constants[index] = value
							end
						else
							for _index, _value in pairs(value) do
								if module.Constants[index][_index] == nil then
									module.Constants[index][_index] = _value
								end
							end
						end
					end
				end
			end
			
			for index, value in pairs(component.Properties) do
				if type(index) == "string" then
					if module.Properties[index] == nil  then
						if value == defin.Nil then
							value = nil
						end
						module.Properties[index] = value
					elseif type(value) == "table" and type(module.Properties[index]) == "table" then
						if isArray(value) and isArray(module.Properties[index]) then
							if not module.Properties[index] then
								module.Properties[index] = value
							end
						else
							for _index, _value in pairs(value) do
								if module.Properties[index][_index] == nil then
									module.Properties[index][_index] = _value
								end
							end
						end
					end
				end
			end
			
			for index, value in pairs(component) do
				if type(value) == "function" and not module[index] and index ~= "Initialized" and index ~= "Destroyed" then
					module[index] = value
				end
			end

			for index, value in pairs(component.Thread) do
				if type(value) == "function" and not module.Thread[index] then
					module.Thread[index] = value
				end
			end
			
			for index, value in pairs(component[context]) do
				if type(value) == "function" and not module[context][index] and index ~= "Initialized" and index ~= "Destroyed" then
					module[context][index] = value
				end
			end

			for index, value in pairs(component[context].Thread) do
				if type(value) == "function" and not module[context].Thread[index] then
					module[context].Thread[index] = value
				end
			end
			
			if type(component.Initialized) == "function" then
				module.InitializedFunctions[componentName] = component.Initialized
			end
			
			if type(component[context].Initialized) == "function" then
				module.InitializedFunctions[componentName] = component[context].Initialized
			end
			
			for index, value in pairs(component.Changed) do
				if type(value) == "function" then
					if not module.ChangedFunctions[index] then
						module.ChangedFunctions[index] = {}
					end
					module.ChangedFunctions[index][componentName] = value
				end
			end
			
			for index, value in pairs(component[context].Changed) do
				if type(value) == "function" then
					if not module.ChangedFunctions[index] then
						module.ChangedFunctions[index] = {}
					end
					module.ChangedFunctions[index][componentName] = value
				end
			end
			
			for index, value in pairs(component.Server.Request) do
				if type(value) == "function" and not module.Server.Request[index] then
					module.Server.Request[index] = value
				end
			end
			
			for index, value in pairs(component.Client.Replicate) do
				if type(value) == "function" and not module.Client.Replicate[index] then
					module.Client.Replicate[index] = value
				end
			end
			
			if type(component.Destroyed) == "function" then
				module.DestroyedFunctions[componentName] = component.Destroyed
			end
			
			if type(component[context].Destroyed) == "function" then
				module.DestroyedFunctions[componentName] = component[context].Destroyed
			end
		end
	
		module.PropertyIndexes = {}
		module.ConstantIndexes = {}
	
		local _index = 5
		
		for index, value in pairs(module.Constants) do
			if type(index) == "string" then
				module.ConstantIndexes[index] = tostring(_index)
			else
				error('Constant index "' .. tostring(index) .. '" must be a string (' .. moduleName .. ')')
			end

			_index = _index + 1
		end

		_index = 5

		for index, value in pairs(module.Properties) do
			if type(index) == "string" then
				module.PropertyIndexes[index] = tostring(_index)
			else
				error('Property index "' .. tostring(index) .. '" must be a string (' .. moduleName .. ')')
			end
			
			_index = _index + 1
		end

		for index, value in pairs(module[context]) do
			if type(value) == "function" then
				module[index] = value
			end
		end
		
		for index, value in pairs(module[context].Thread) do
			if type(value) == "function" then
				module.Thread[index] = value
			end
		end
		
		if type(module.Initialized) == "function" then
			module.InitializedFunctions.Class = module.Initialized
		end
		
		if type(module[context].Initialized) == "function" then
			module.InitializedFunctions.Class = module[context].Initialized
		end
		
		for index, value in pairs(module.Changed) do
			if type(value) == "function" then
				if not module.ChangedFunctions[index] then
					module.ChangedFunctions[index] = {}
				end
				module.ChangedFunctions[index].Class = value
			end
		end
		
		for index, value in pairs(module[context].Changed) do
			if type(value) == "function" then
				if not module.ChangedFunctions[index] then
					module.ChangedFunctions[index] = {}
				end
				module.ChangedFunctions[index].Class = value
			end
		end
		
		if type(module.Destroyed) == "function" then
			module.DestroyedFunctions.Class = module.Destroyed
		end
		
		if type(module[context].Destroyed) == "function" then
			module.DestroyedFunctions.Class = module[context].Destroyed
		end
		
		module.Request = module.Server.Request
		module.Replicate = module.Client.Replicate
		
		if isClient then
			for index in pairs(module.Request) do
				module["Request" .. index] = function(instance, ...)
					return eventsFolder.InstanceRequestFunction:InvokeServer(instance.Container, index, unpack(defin:UnwrapTable({ ... })))
				end
			end
		else
			for index in pairs(module.Replicate) do
				module["Replicate" .. index] = function(instance, ...)
					if canReplicateInstance(instance) then
						eventsFolder.InstanceReplicateFunction:FireAllClients(instance["1"], index, unpack(defin:UnwrapTable({ ... })))
					elseif isInstanceReplicationPaused(instance) then
						queueInstanceReplicationUpdate(instance, {
							2,
							false,
							index,
							defin:UnwrapTable({ ... })
						})
					end
				end
				
				module["ReplicatePlayer" .. index] = function(instance, player, ...)
					if canReplicateInstance(instance) then
						eventsFolder.InstanceReplicateFunction:FireClient(defin:Unwrap(player), instance["1"], index, unpack(defin:UnwrapTable({ ... })))
					elseif isInstanceReplicationPaused(instance) then
						queueInstanceReplicationUpdate(instance, {
							2,
							defin:Unwrap(player),
							index,
							defin:UnwrapTable({ ... })
						})
					end
				end
			end
		end
		
		if moduleType == "Class" then
			function module:Clone()
				return defin:NewInstance(defin:CloneValue(self))
			end

			function module:SetClass(className)
				if classes[className] then
					rawset(self, "2", className)

					if isServer then
						eventsFolder.InstanceClassChanged:FireAllClients(self.Container, className)
					end
				end
			end

			function module:StartThread(name)
				startInstanceThread(self, name)
			end

			function module:StopThread(name)
				stopInstanceThread(self, name)
			end

			function module:IsThreadActive(name)
				isInstanceThreadActive(self, name)
			end
		end

		function module:IsA(_className)
			return self.ClassName == _className or _className == "DefinInstance" or self.Container:IsA(_className) or table.find(module.Components, _className) ~= nil
		end

		function module:Debug(...)
			if DEBUG then
				print(self, ...)
			end
		end
	end
end

function setupContainers()
	local physicalContainers = workspace:FindFirstChild("PhysicalContainers")

	if physicalContainers then
		for _, container in pairs(physicalContainers:GetChildren()) do
			container.Parent = replicatedStorage.Containers
		end

		physicalContainers:Destroy()
	end

	for _, container in pairs(replicatedStorage.Containers:GetChildren()) do
		containers[container.Name] = container
	end
end

function setupModules()
	for moduleName, module in pairs(components) do
		setupModule(moduleName, "Component", module)
	end
	
	for moduleName, module in pairs(classes) do
		setupModule(moduleName, "Class", module)
	end

	for moduleName, module in pairs(services) do
		setupModule(moduleName, "Service", module)
	end
end

function requireModules()
	for _, module in pairs(script:GetDescendants()) do
		if module:IsA("ModuleScript") then
			local moduleType = module:FindFirstAncestorOfClass("Configuration")
			moduleType = moduleType and moduleType.Name
			
			local _table = tablesForModuleTypes[moduleType]

			if _table then
				requireModule(_table, module)
			end
		end
	end
end

function hideInstance(instance)
	if instance then
		instance.Name = ""
		instance.Parent = createFolder("", createFolder("", createFolder("")))
		
		return instance
	end
end

function createFolder(name, parent)
	local folder
	
	if parent then
		folder = parent:FindFirstChild(name)
	end
	
	if not folder then
		folder = Instance.new("Folder")
		folder.Name = name
		folder.Parent = parent
	end
	
	return folder
end

function combineTables(table1, table2)
	for index, value in pairs(table1) do
		if type(value) == "table" and type(table2[index]) == "table" then
			combineTables(value, table2[index])
		else
			table2[index] = value
		end
	end
	
	return table2
end

function isArray(_table)
	if type(_table) == "table" then
		local amount = 0
		
		for _ in pairs(_table) do
			amount = amount + 1
		end
		
		return amount > 0 and #_table == amount
	end
end

function isInstanceReplicated(instance)
	return instance["4"]
end

function isInstanceReplicationPaused(instance)
	return pausedInstanceReplications[instance] ~= nil
end

function canReplicateInstance(instance)
	return isServer and isInstanceReplicated(instance) and not isInstanceReplicationPaused(instance)
end

function replicateInstance(instance)
	if instance.Parent and not instance:IsDescendantOf(serverStorage) and not instance:IsDescendantOf(serverScriptService) and instance.Parent ~= game and classes[instance.ClassName] then
		instance["4"] = true
		eventsFolder.InstanceAdded:FireAllClients(instance)
	end
end

function checkReplicationBundle(instance)
	for class, _instances in pairs(replicationBundles) do
		if instance:IsA(class) then
			table.insert(_instances, instance)
			return true
		end
	end
	
	return false
end

function getModulesAmount()
	return #script.Classes:GetChildren() + #script.Services:GetChildren() + #script.Components:GetChildren()
end

function checkPlayerRequest(player, requestName)
	if playerRequests[player.UserId][requestName] then
		player:Kick("D1")
		
		return false
	else
		playerRequests[player.UserId][requestName] = tick()
		
		return true
	end
end

function endClient()
	while true do	end
end

function initialize()
	-- Set context

	if isClient then
		context = "Client"
		
		if not isStudio then
			hideInstance(script)

			local modulesAmount = getModulesAmount()
			local definIndexes = 0

			setmetatable(_G, {
				__metatable = DEFIN_LOCKED,
				
				__index = function(_, index)
					if index == "Defin" then
						if definIndexes < modulesAmount then
							definIndexes = definIndexes + 1
							
							return defin
						else
							endClient()
						end
					end
				end
			})
		end

		eventsFolder = replicatedStorage.DefinEvents

		eventsFolder.InstancePropertyFunction.OnClientEvent:Connect(function(instanceContainer, propertyName, newValue)
			if instanceContainer then
				local instance = defin:WaitForWrap(instanceContainer, 10)

				if defin:IsWrapped(instance) then
					changeInstanceProperty(instance, propertyName, newValue, false, true)
				end
			end
		end)

		eventsFolder.InstanceReplicateFunction.OnClientEvent:Connect(function(instanceContainer, functionName, ...)
			if instanceContainer then
				local instance = defin:WaitForWrap(instanceContainer, 10)
				
				if defin:IsWrapped(instance) then
					local class = getClassForInstance(instance)
					local func = class.Replicate[functionName]

					func(instance, unpack(defin:WrapTable({ ... }, true)))
				end
			end
		end)
	elseif isServer then
		context = "Server"

		_G.DefinVersion = 181

		eventsFolder = createFolder("DefinEvents", replicatedStorage)
		
		Instance.new("RemoteFunction", eventsFolder).Name = "GetServiceInstances"
		Instance.new("RemoteFunction", eventsFolder).Name = "GetClassInstances"
		Instance.new("RemoteEvent", eventsFolder).Name = "InstanceAdded"
		Instance.new("RemoteEvent", eventsFolder).Name = "InstancesAdded"
		Instance.new("RemoteEvent", eventsFolder).Name = "InstanceReplicationContinued"
		Instance.new("RemoteEvent", eventsFolder).Name = "InstanceClassChanged"
		Instance.new("RemoteEvent", eventsFolder).Name = "InstancePropertyFunction"
		Instance.new("RemoteFunction", eventsFolder).Name = "InstanceRequestFunction"
		Instance.new("RemoteEvent", eventsFolder).Name = "InstanceReplicateFunction"

		eventsFolder.InstanceRequestFunction.OnServerInvoke = function(player, instanceContainer, functionName, ...)
			local instance = defin:Wrap(instanceContainer)
			
			if defin:IsWrapped(instance) then
				local class = getClassForInstance(instance)
				local func = class.Request[functionName]
				
				if type(func) == "function" then
					return func(instance, defin:Wrap(player), unpack(defin:WrapTable({ ... })))
				end
			end
		end
	end

	_G.Defin = defin
	
	-- Setup modules

	setupContainers()
	requireModules()
	setupModules()

	defin.Content = replicatedStorage:WaitForChild("Content")

	if isClient then
		if not isStudio then
			_G.Defin = nil
		end
		
		for _, instance in pairs(eventsFolder.GetServiceInstances:InvokeServer()) do
			task.spawn(function()
				serviceInstances[instance["2"]] = defin:NewInstance(instance)
			end)
		end
	
		for _, service in pairs(serviceInstances) do
			defin:WaitForWrap(service)
		end
		
		eventsFolder.InstanceAdded.OnClientEvent:Connect(function(instance)
			if instance["1"] then
				defin:NewInstance(instance)
			end
		end)
		
		eventsFolder.InstancesAdded.OnClientEvent:Connect(function(instances)
			for _, instance in pairs(instances) do
				if instance["1"] then
					task.spawn(function()
						defin:NewInstance(instance)
					end)
				end
			end
		end)

		eventsFolder.InstanceReplicationContinued.OnClientEvent:Connect(function(instanceContainer, updates)
			local instance = instances[instanceContainer]

			if instance then
				local class = getClassForInstance(instance)

				for _, update in pairs(updates) do
					local _type = update[1]

					if _type == 1 then
						local propertyName = update[2]
						local propertyValue = update[3]

						changeInstanceProperty(instance, propertyName, propertyValue, false, true)
					elseif _type == 2 then
						local forPlayer = update[2]

						if not forPlayer or forPlayer == localPlayer then
							local functionName = update[3]
							local parameters = update[4]
							local func = class.Replicate[functionName]

							func(instance, unpack(defin:WrapTable(parameters, true)))
						end
					end
				end
			end
		end)

		for _, instance in pairs(eventsFolder.GetClassInstances:InvokeServer()) do
			if instance["1"] then
				task.spawn(function()
					defin:NewInstance(instance)
				end)
			end
		end
		
		eventsFolder.InstanceClassChanged.OnClientEvent:Connect(function(instanceContainer, className)
			local instance = instances[instanceContainer]

			if instance then
				instance:SetClass(className)
			end
		end)

		if not isStudio then
			eventsFolder.GetServiceInstances:Destroy()
			eventsFolder.GetClassInstances:Destroy()
			hideInstance(eventsFolder.InstanceAdded)
			hideInstance(eventsFolder.InstancesAdded)
			hideInstance(replicatedStorage.Content)
			hideInstance(replicatedStorage.Containers)
		end
	elseif isServer then
		-- Return all existing instances when invoked by a client
		
		eventsFolder.GetServiceInstances.OnServerInvoke = function(player)
			if checkPlayerRequest(player, "GetServiceInstances") then
				playerRequests[player.UserId].GetServiceInstances = tick()
				local _instances = {}
				
				for index, _ in pairs(services) do
					local service = defin:GetService(index)
					table.insert(_instances, service)
				end
				
				return _instances
			end
		end

		eventsFolder.GetClassInstances.OnServerInvoke = function(player)
			if checkPlayerRequest(player, "GetClassInstances") then
				local _instances = {}
				
				for _, instance in pairs(instances) do
					if classes[instance.ClassName] then
						table.insert(_instances, instance)
					end
				end
				
				return _instances
			end
		end

		-- Instance and index to serviceInstances all game services

		for index, _ in pairs(services) do
			task.spawn(function()
				serviceInstances[index] = defin:NewInstance(index, replicatedStorage)
			end)
		end

		-- Instance player class on join
		
		players.PlayerAdded:Connect(function(player)
			playerRequests[player.UserId] = {}

			if classes.Player then
				defin:NewInstance("Player", nil, player)
			end
		end)
		
		players.PlayerRemoving:Connect(function(player)
			playerRequests[player.UserId] = nil
		end)
	end
end

initialize()

-- Begin Types

-- ========================================
-- COMPONENTS
-- ========================================

-- ----------------------------------------
-- Module: TestComponentA
-- Source: scripts/Components/TestComponentA.luau
-- ----------------------------------------

local TestComponentAConstants = {
	ComponentConstant1 = "Hello",
	ComponentConstant2 = 123,
	ComponentConstant3 = true,
	ComponentConstant4 = {
		Foo = "Bar",
		Baz = 456,
		Qux = false,
	}
}

export type TestComponentAConstants = typeof(TestComponentAConstants)

local TestComponentAProperties = {
	ComponentProperty1 = "World",
	ComponentProperty2 = 456,
	ComponentProperty3 = false,
	ComponentProperty4 = nil :: Player?
}

export type TestComponentAProperties = typeof(TestComponentAProperties)

export type TestComponentAMethods = {
	ComponentAMethod: ((self: TestComponentA, arg1: string, arg2: number) -> Player?),
}

export type TestComponentAClientMethods = {
	ComponentAClientMethod: ((self: TestComponentAClient, arg1: string) -> any),
	ComponentAMethod: ((self: TestComponentAClient, arg1: string, arg2: number) -> Player?),
	Initialized: ((self: TestComponentAClient) -> any),
	RequestComponentAServerMethod: ((self: TestComponentAClient) -> any),
}

export type TestComponentAServerMethods = {
	ComponentAMethod: ((self: TestComponentAServer, arg1: string, arg2: number) -> Player?),
	ComponentAServerMethod: ((self: TestComponentAServer, arg1: string, arg2: number) -> any),
	Initialized: ((self: TestComponentAServer) -> any),
	ReplicateComponentAClientMethod: ((self: TestComponentAServer, arg1: string) -> any),
	ReplicatePlayerComponentAClientMethod: ((self: TestComponentAServer, player: Player, arg1: string) -> any),
}

export type TestComponentA = TestComponentAConstants & TestComponentAProperties & TestComponentAMethods & Folder

export type TestComponentAClient = TestComponentA & TestComponentAClientMethods

export type TestComponentAServer = TestComponentA & TestComponentAServerMethods

-- ========================================
-- CLASSES
-- ========================================

-- ----------------------------------------
-- Module: TestClass
-- Source: scripts/Classes/TestClass.luau
-- ----------------------------------------

local TestClassConstants = {
	ClassConstant1 = "Hello",
	ClassConstant2 = 123,
	ClassConstant3 = true,
	ClassConstant4 = {
		Foo = "Bar",
		Baz = 456,
		Qux = false,
	}
}

export type TestClassConstants = typeof(TestClassConstants) & TestComponentAConstants

local TestClassProperties = {
	ClassProperty1 = "World",
	ClassProperty2 = 456,
	ClassProperty3 = false,
	ClassProperty4 = nil :: Player?
}

export type TestClassProperties = typeof(TestClassProperties) & TestComponentAProperties

export type TestClassMethods = {
	ClassMethod: ((self: TestClass, arg1: string, arg2: number) -> Player?),
} & TestComponentAMethods

export type TestClassClientMethods = {
	ClassClientMethod: ((self: TestClassClient, arg1: string) -> any),
	ClassMethod: ((self: TestClassClient, arg1: string, arg2: number) -> Player?),
	Initialized: ((self: TestClassClient) -> any),
	RequestClassServerMethod: ((self: TestClassClient) -> any),
} & TestComponentAClientMethods

export type TestClassServerMethods = {
	ClassMethod: ((self: TestClassServer, arg1: string, arg2: number) -> Player?),
	ClassServerMethod: ((self: TestClassServer, arg1: string, arg2: number) -> any),
	Initialized: ((self: TestClassServer) -> any),
	ReplicateClassClientMethod: ((self: TestClassServer, arg1: string) -> any),
	ReplicatePlayerClassClientMethod: ((self: TestClassServer, player: Player, arg1: string) -> any),
} & TestComponentAServerMethods

export type TestClass = TestClassConstants & TestClassProperties & TestClassMethods & Model

export type TestClassClient = TestClass & TestClassClientMethods

export type TestClassServer = TestClass & TestClassServerMethods

-- End Types

return defin